From a938f4b392dc804fbd538e306fa2593b9395dc50 Mon Sep 17 00:00:00 2001
From: Dave Stevenson <dave.stevenson@raspberrypi.org>
Date: Mon, 12 Feb 2018 17:25:05 +0000
Subject: [PATCH 12/80] staging: bcm2835-v4l2: Convert to using device tree

Switch from being a magic module to loading via device
tree, and hence getting a platform_device handle.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>
---
 .../bcm2835-camera/bcm2835-camera.c           | 33 ++++++++++++++++---
 .../bcm2835-camera/bcm2835-camera.h           |  1 +
 2 files changed, 30 insertions(+), 4 deletions(-)

diff --git a/drivers/staging/vc04_services/bcm2835-camera/bcm2835-camera.c b/drivers/staging/vc04_services/bcm2835-camera/bcm2835-camera.c
index 9e6d62184452..00bcd2a4dbe3 100644
--- a/drivers/staging/vc04_services/bcm2835-camera/bcm2835-camera.c
+++ b/drivers/staging/vc04_services/bcm2835-camera/bcm2835-camera.c
@@ -16,6 +16,7 @@
 #include <linux/errno.h>
 #include <linux/kernel.h>
 #include <linux/module.h>
+#include <linux/platform_device.h>
 #include <linux/slab.h>
 #include <media/videobuf2-vmalloc.h>
 #include <media/videobuf2-dma-contig.h>
@@ -1859,7 +1860,7 @@ static struct v4l2_format default_v4l2_format = {
 	.fmt.pix.sizeimage = 1024 * 768,
 };
 
-static int __init bm2835_mmal_init(void)
+static int bm2835_mmal_probe(struct platform_device *pdev)
 {
 	int ret;
 	struct bm2835_mmal_dev *dev;
@@ -1887,6 +1888,7 @@ static int __init bm2835_mmal_init(void)
 			goto cleanup_gdev;
 		}
 
+		dev->pdev = pdev;
 		dev->camera_num = camera;
 		dev->max_width = resolutions[camera][0];
 		dev->max_height = resolutions[camera][1];
@@ -1931,6 +1933,7 @@ static int __init bm2835_mmal_init(void)
 		q->ops = &bm2835_mmal_video_qops;
 		q->mem_ops = &vb2_vmalloc_memops;
 		q->timestamp_flags = V4L2_BUF_FLAG_TIMESTAMP_MONOTONIC;
+		q->dev	= &pdev->dev;
 		ret = vb2_queue_init(q);
 		if (ret < 0)
 			goto unreg_dev;
@@ -1979,7 +1982,7 @@ static int __init bm2835_mmal_init(void)
 	return ret;
 }
 
-static void __exit bm2835_mmal_exit(void)
+static int bm2835_mmal_exit(struct platform_device *pdev)
 {
 	int camera;
 	struct vchiq_mmal_instance *instance = gdev[0]->instance;
@@ -1989,7 +1992,29 @@ static void __exit bm2835_mmal_exit(void)
 		gdev[camera] = NULL;
 	}
 	vchiq_mmal_finalise(instance);
+
+	return 0;
 }
 
-module_init(bm2835_mmal_init);
-module_exit(bm2835_mmal_exit);
+/*
+ *   Register the driver with device tree
+ */
+
+static const struct of_device_id bcm2835_v4l2_of_match[] = {
+	{.compatible = "raspberrypi,bcm2835-v4l2",},
+	{ /* sentinel */ },
+};
+
+MODULE_DEVICE_TABLE(of, bcm2835_v4l2_of_match);
+
+static struct platform_driver bcm2835_v4l2_driver = {
+	.probe = bm2835_mmal_probe,
+	.remove = bm2835_mmal_exit,
+	.driver = {
+		   .name = BM2835_MMAL_MODULE_NAME,
+		   .owner = THIS_MODULE,
+		   .of_match_table = bcm2835_v4l2_of_match,
+		   },
+};
+
+module_platform_driver(bcm2835_v4l2_driver);
diff --git a/drivers/staging/vc04_services/bcm2835-camera/bcm2835-camera.h b/drivers/staging/vc04_services/bcm2835-camera/bcm2835-camera.h
index 31333f57c602..35b4ffd0759f 100644
--- a/drivers/staging/vc04_services/bcm2835-camera/bcm2835-camera.h
+++ b/drivers/staging/vc04_services/bcm2835-camera/bcm2835-camera.h
@@ -37,6 +37,7 @@ enum {
 extern int bcm2835_v4l2_debug;
 
 struct bm2835_mmal_dev {
+	struct platform_device *pdev;
 	/* v4l2 devices */
 	struct v4l2_device     v4l2_dev;
 	struct video_device    vdev;
-- 
2.17.0

